src/main/java/bio/terra/stairway/HookWrapper.java

```diff
@@ -66,34 +79,48 @@ void checkHookAction(HookAction action) {
     this.stairwayHooks = stairwayHooks;
   }
 
-  void startFlight(FlightContext flightContext) {
+  void startFlight(FlightContext flightContext) throws InterruptedException {
+    // First handle plain flight hooks
     handleHookList(flightContext, HookOperation.START_FLIGHT);
+    // Use the factory to collect any flight hooks for this flight
+    List<DynamicHook> flightHooks = new ArrayList<>();
+    for (StairwayHook stairwayHook : stairwayHooks) {
+      Optional<DynamicHook> maybeFlightHook = stairwayHook.flightFactory(flightContext);
+      maybeFlightHook.ifPresent(flightHooks::add);
+    }
+    // Then handle any flight hooks list from the factory
+    handleDynamicHookList(flightHooks, flightContext, HookOperation.START_FLIGHT);
+    flightContext.setFlightHooks(flightHooks);
   }
 
   void startStep(FlightContext flightContext) throws InterruptedException {
     // First handle plain step hooks
     handleHookList(flightContext, HookOperation.START_STEP);
     // Use the factory to collect any step hooks for this step
-    List<StepHook> stepHooks = new ArrayList<>();
+    List<DynamicHook> stepHooks = new ArrayList<>();
     for (StairwayHook stairwayHook : stairwayHooks) {
-      Optional<StepHook> maybeStepHook = stairwayHook.stepFactory(flightContext);
+      Optional<DynamicHook> maybeStepHook = stairwayHook.stepFactory(flightContext);
       maybeStepHook.ifPresent(stepHooks::add);
     }
     // Then handle any step hooks list from the factory
-    handleStepHookList(stepHooks, flightContext, HookOperation.START_STEP);
+    handleDynamicHookList(stepHooks, flightContext, HookOperation.START_STEP);
     flightContext.setStepHooks(stepHooks);
   }
 
   void endStep(FlightContext flightContext) {
     // First handle plain step hooks
     handleHookList(flightContext, HookOperation.END_STEP);
     // Next handle the step hooks list from the flight context
-    handleStepHookList(flightContext.getStepHooks(), flightContext, HookOperation.END_STEP);
+    handleDynamicHookList(flightContext.getStepHooks(), flightContext, HookOperation.END_STEP);
     flightContext.setStepHooks(null);
   }
 
   void endFlight(FlightContext flightContext) {
+    // First handle plain flight hooks
     handleHookList(flightContext, HookOperation.END_FLIGHT);
+    // Next handle the flight hooks list from the flight context
+    handleDynamicHookList(flightContext.getFlightHooks(), flightContext, HookOperation.END_FLIGHT);
+    flightContext.setFlightHooks(null);
```

Comment:
it's a little unusual to me that the flight hooks are removed from the FlightContext, but I don't have much of a reason for keeping them, just a possible bug if they get examined after `endFlight`. I wouldn't particularly expect `HookWrapper#endFlight` to modify the FlightContext this way, though it does match `startFlight`. Any particular reason for removing them?

Reply:
No strong reason. It was symmetric with endStep logic. I wouldn't mind an NPE if some code assumed that list was always there when it was no longer valid.